### ISSUE TITLE
compiling ospTutorial on Cray

### ISSUE THREAD

**Issue by dkoerner** [ospray/OSPRay#77]
Hi,
I successfully compiled the ospray library and its dependencies on a cray system using the cray wrapper to gcc. However, the build process fails when trying to compile the ospTutorial application due to the following linker error:
/usr/bin/ld: attempted static link of dynamic object `../../../libospray.so.0.10.1'

I tracked the problem down to the fact that for some obscure reason the cray wrapper does not allow linking against a mix of static and dynamic libraries. Either all inputs have to be static, or dynamic but they can't be mixed.

To work around that problem I tried to change the cmake files such that all libraries are build as dynamic libraries. This was already the case for most of the targets and only some embree targets had to be changed.

The problem now is that when I compile ospTutorial using only dynamic libraries, I get a bunch of other linker errors of the type:
../../../libembree.so.2: undefined reference to `embree::avx::BVH4GridAOSIntersector8'
[...]

I am linking to the following embree isa-targets:
libembree_sse42.so
libembree_avx.so
libembree_avx2.so

I tried to get to the root of the probem but I am too unfamiliar with the way embree is set up. I have the feeling that the problem is related to the fact that the cray wrapper always has __AVX2__ defined by default while g++ hasn't. This probably messes up the different control paths used to compile embree for the different ISA-targets. It is interesting that the linker complains about missing symbols from embree::avx and embree::sse42 namespaces but not the embree::avx2 namespace.

So my question here is:
Which compiler to you use for building ospTutorial/ospRay on the Cray? If you also use the gcc wrapper provided by Cray, how do you deal with the problem of not being allowed to mix static/dynamic libraries?

Best,
David

**Comment by carsonbrownlee** [comment]
hi David,
I've only compiled on one Cray system before at TACC.  I'll go dig back and check what I used for that.

**Comment by carsonbrownlee** [comment]
I compiled with 
icc version 16.0.1 (gcc version 4.9.3 compatibility)

**Comment by carsonbrownlee** [comment]
I get the same error with gcc, even when explicitly setting OSPRAY_BUILD_ISA.  
The problem seems to be in the generated isa target libraries

nm libembree_sse42.a | grep BVH4Triangle4Intersector4HybridMoeller | c++filt
0000000000000760 B embree::avx::BVH4Triangle4Intersector4HybridMoeller
0000000000000740 B embree::avx::BVH4Triangle4Intersector4HybridMoellerNoFilter

that should be embree::sse42.  Never seen this before, does not show up with ICC.  Jeff may have an easy answer for this.

**Comment by johguenther** [comment]
One workaround would be to use a [pre-compiled Embree](https://github.com/embree/embree/releases), i.e. enable `OSPRAY_USE_EXTERNAL_EMBREE` (and maybe set `embree_DIR`) in CMake. The different ISA-specific sub-libraries of Embree *need* to be compiled with the matching ISA flags, otherwise the namespaces get mixed up, causing liker errors (you mentioned the Cray wrappers always use AVX2).

**Comment by dkoerner** [comment]
Hi Johannes,
thanks for the suggestion. I now compiled embree v2.7.1 externally using just the intel compiler. Then I built ospray using the OSPRAY_USE_EXTERNAL_EMBREE and the cray wrapper around the intel compiler. This worked fine.

Now when I try to compile ospTutorial, I get some undefined reference errors again which look like this:
../../../libospray.so.1.0.0: undefined reference to `rtcDeleteScene___un_3C_s_5B_un__RTCScene_5D__3E_avx512knl'
 [...]

So the original problem went away by compiling embree externally without cray wrapper but it seems that the wrapper still breaks something when compiling ospray itsself. It seems to be related to the avx512knl ISA target since all undefined references have that in their name.

Is there a way to turn of avx512knl?

Thanks for all your comments so far!

**Comment by johguenther** [comment]
Yes, it seems that Embree was compiled without AVX-512 support. To limit OSPRay to only use AVX2 set `OSPRAY_BUILD_ISA` to `AVX2`, then it should link. Alternatively you can also download Embree with AVX-512 from the link above (v2.10 is compatible).

**Comment by ingowald** [comment]
The AVX2 being _always_ defined is probably the issue here; embree will 
internally build kernels for different ISAs (say 4-wide on SSE but 
8-wide on AVX); it does so by compiling different static libraries for 
sse4, avx, avx2, etc, and linking those together at the end - while in 
each library it uses the compiler flags to figure out which kernels to 
include (with which flags) into this library - thus, if the compiler 
always turns on the avx2 flag even if embree explicitly compiles the 
sse4 library with sse4 compile flags then this sse4 library will likely 
include some wrong kernels (since the compiler tells it it was actually 
avx2). Eg, in the example you mention it didn't put the 
bvh4gridaosintersector8 into the avx module because internally checked 
the 'AVX2' flag, and then decided that since it's avx2 it should rather 
use the bvh*8* intersectors (pretty sure you'll have the same problem 
for the sse4 namespace, actually).

Not 100% sure that's the problem, but almost. Of course, the root cause 
of the problem is that the compiler is actually wrong here: it should 
_not_ specify the AVX2 flag when building an sse or avx1 target. Do you 
have access to a version of the intel compiler? Does it work with this?

Otherwise, the only coarse of action we have is to add some 
"WORKAROUND_GCC_AVX2_BUG=1" compile flags to the sse4 and avx1 
libraries, then add some
#ifdef WORKAROUND_GCC_AVX2_BUG
#  undef AVX2
#endif
or something into some commonly included header file.

Cheers,

Ingo

**Comment by carsonbrownlee** [comment]
Could you try the same compiler version I used above to build on cray to see if this goes away, compiling embree internally to OSPRay?  icc version 16.0.1 (gcc version 4.9.3 compatibility).

**Comment by ingowald** [comment]
In the cmake dialog you can set the build ISA to whatever ISA you want.

The problem you are seeing is that the default value for this field is 
"ALL", and since ospray since newest version supports KNL it will try to 
build for all ISAs _including_ KNL, even though the embree you build 
against was compiled without KNL support.

I already have a gitlab issue open for that (I fell into that trap 
myself), and we'll very soon change the "ALL" field to mean "all targets 
that are also supported by your embree version" - but for now, just set 
the ISA to "AVX2", and the problem should go away.

Sorry for the inconvenience; you stumbled over this bug faster than we 
could fix it ;-)

Cheers,

Ingo

**Comment by dkoerner** [comment]
Hi,
sorry for the delayed reply. With the help of your comments and some additional hints from some guys at cray I got it to work. Here is a summary of the things I had to do:

-Note that all this applies only if you want to build ospray with MPI support on the cray using -DOSPRAY_BUILD_MPI_DEVICE=ON

-Embree needs to be built externally without the cray wrapper. So when building Embree, the compiler has to be set explicitly via CMAKE_C_COMPILER/CMAKE_CXX_COMPILER flags in order to not have cmake pick up the wrapper. This can be icc/icpc or gcc/g++. The reason for that is that the cray wrapper adds an isa target flag which breaks the build for the different isa targets in embree.

-Now for building ospray, the cray wrappers to the intel or gnu compiler should be used (cc/CC) in order to get proper MPI support. Now the problem is, that FindMPI.cmake will try a bunch of things to figure out the MPI library flags and only if everything fails, it will, as a last resort, try to see if the used compiler (in this case the cray wrapper) actually is able to compile mpi code on its own. If thats the case, cmake will set all flags accordingly and carry on. This means that if there happens to be an mpiexec in the path, cmake will interrogate it first and use the flags coming from there. This can lead to the case that cmake will use openmpi flags (coming from mpiexec -showme) and passes those along to the compiler, which itsself will pass on its own mpi flags (e.g. mpich) if the compiler happens to be an mpi wrapper. Obviously this can cause different kind of issues further down the road. Now to prevent this from happening, I made a custom FindMPI.cmake version in ospray.git/cmake. I turned of compiler interrogation alltogether by commenting out the call to interrogate_mpi_compiler near the end of the file. This causes a call later to try_regular_compiler which will succeed if a wrapper is used. A second option could be to try to get rid of any mpi compilers (such as mpicc etc.) and all other mpi executeables (such as mpiexec, mpirun, srun, etc.) in the path. However, this was not possible in my cray environment. A third option could be to not use the wrapper and make sure that executeables such as mpiexec return flags for the mpi library which ospray is supposed to be using. From what I can tell this is what you guys have been doing.

-As mentioned above, when building ospray, the use of an external build of embree has to be indicated via OSPRAY_USE_EXTERNAL_EMBREE=ON (and emree_DIR, if embree doesnt reside in CMAKE_INSTALL_PREFIX).

-As mentioned, using the wrapper will add some isa target flag (for avx2 in our case). This still leads to problems with symbol references from different isa targets when linking with embree. Therefore it is necessary to explicitly tell ospray which isa target to use. We do this by using -DOSPRAY_BUILD_ISA=AVX2

-Another issue with the cray wrapper is, that when building executeables, it requires all dependencies to be linked statically or dynamically. So there is no mixing of static or shared dependencies allowed. By default the wrapper assumes that all dependecies have to be static which causes compiler errors during build. We need to expliticly tell the wrapper that all the dependencies will be shared libraries. This is done by passing a "-dynamic" flag to the cray wrapper using -DCMAKE_CXX_FLAGS=-dynamic -DCMAKE_C_FLAGS=-dynamic

-note: When building for the cray, we don't want to build any of the targets which use opengl or qt. So we turn those of by doing:
-DOSPRAY_APPS_QTVIEWER=OFF
-DOSPRAY_APPS_VOLUMEVIEWER=OFF
-DOSPRAY_APPS_GLUTVIEWER=OFF
-DOSPRAY_MODULE_OPENGL_UTIL=OFF

Thanks alot for your help so far!

Best,
David
